File: workshop_2/python/lin_reg.py
import numpy as np
import pandas as pd

x_lin = pd.read_csv("workshop_2/data/x_create_linear_data.csv",header=None)
y_lin = pd.read_csv("workshop_2/data/y_create_linear_data.csv",header=None)

# append ones column to find intercept
# https://stackoverflow.com/questions/45061326/how-we-can-compute-intercept-and-slope-in-statsmodels-ols
x_lin = pd.concat([x_lin, pd.DataFrame(np.ones_like(x_lin[0]))], axis=1)

# https://jianghaochu.github.io/ordinary-least-squares-regression-in-python-from-scratch.html
beta_hat = np.array(np.matmul(
    np.matmul(np.linalg.inv(np.matmul(np.array(x_lin).transpose(), 
    np.array(x_lin))), 
    x_lin.transpose()), 
    y_lin
))

def predict(x_new,beta_hat):
    pred = 0
    for i in range(0,len(x_new)):
        pred += x_new[i] * beta_hat[i]
    pred += beta_hat[-1]
    return pred

print("Coefficients: ",beta_hat[:-1])
print("Intercept: ",beta_hat[-1]) 

x_new = [30,40,50]
print("Prediction: ",predict(x_new,beta_hat))
print("Correct: ",sum(x_new))

File: workshop_2/R/lin_reg.R
x_linear <- read.csv("workshop_2/data/x_create_linear_data.csv", header = FALSE)
y_linear <- read.csv("workshop_2/data/y_create_linear_data.csv", header = FALSE)

colnames(y_linear)[1] <- "Dependent_Var"
colnames(x_linear)[1] <- "Covariate_A"
colnames(x_linear)[2] <- "Covariate_B"
colnames(x_linear)[3] <- "Covariate_C"

df <- data.frame(y_linear, x_linear)
model <- lm(Dependent_Var ~ Covariate_A + Covariate_B + Covariate_C, data = df)

if (!dir.exists(file.path("outputs"))) {
  dir.create(file.path("outputs"))
}
if (!dir.exists(file.path("outputs/workshop_2"))) {
  dir.create(file.path("outputs/workshop_2"))
}

capture.output(
  summary(model),
  file = "outputs/workshop_2/Linear Regression.txt"
)

new <- data.frame(Covariate_A = 30.0, Covariate_B = 40.0, Covariate_C = 50.0)
pred <- predict(model, new, type = "response")
print(paste("Prediction: ", pred))
print(paste("Correct: ", sum(new)))
